Use Platform::Memory* to allocate things for minMDNS advertisement

andy31415 opened the pull request:
#### Problem
Forcing static allocation makes the code much harder to read and maintain.

 #### Summary of Changes
Switch minMDNS advertiser to use dynamic memory. Created separate methods for allocating QNames in HEAP as one contigoush array of bytes.

mspang commented on src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:
```diff
@@ -165,77 +169,129 @@ class AllInterfaces : public mdns::Minimal::ListenIterator
+    /// allocated memory.
+    void Clear();
+
+    /// Append another responder to the internal replies
```
super nit: These comments keep switching mood.

"Set the query responder to a blank state and free up any allocated memory."

"Appends another responder to the internal replies."

[Note Google style is to always use the latter, like you've done for Clear]

mspang commented on src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:
```diff
@@ -165,77 +169,129 @@ class AllInterfaces : public mdns::Minimal::ListenIterator
+
+    /// Append another responder to the internal replies
+    template <typename ResponderType, typename... Args>
+    QueryResponderSettings AddResponder(Args &&... args)
```
What's the code size impact of this?

andy31415 replied:
Unsure. I assumed inline, so I did not check. It makes existing code a lot more readable, however I have not checked overead in terms of pure code.

I would rather do one at a time: first make it readable & work and then followup with code cleanup if it makes sense. I would expect simple std::forward templates to properly inline. Although this code is addmitedly a bit more complex.

mspang commented on src/lib/mdns/minimal/core/FlatAllocatedQName.h:
```diff
@@ -0,0 +1,93 @@
+
+} // namespace Internal
+
+/// Build a qname inside the given storate
```
storate -> storage

mspang commented on src/lib/mdns/minimal/core/FlatAllocatedQName.h:
```diff
@@ -0,0 +1,93 @@
+/// Example:
+///   malloc(RequiredStorageSize("myhostname", "local"));
+template <typename... Args>
+inline size_t RequiredStorageSize(QNamePart name, Args &&... rest)
```
What's the code size impact of this? Would C style varargs be smaller?

andy31415 replied:
I think it may be inlinable and the function is super small.

Godbolt shows some code: https://godbolt.org/z/5rEeMo but not overly bad (i.e. a bunch of strlen calls is for the most case).

I would prefer to optimize later if applicable.

andy31415 replied:
I have converted this to varargs and the saving is quite small on this piece of code:

```
❯ bloaty examples/all-clusters-app/esp32/build/chip-all-clusters-app.elf -- ~/tmp/chip-all-clusters-app.elf.memory_alloc_base
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  +0.1% +7.65Ki  [ = ]       0    .debug_info
  +0.1% +2.07Ki  [ = ]       0    .debug_line
  +0.2% +1.65Ki  [ = ]       0    .debug_abbrev
  +0.0%     +64  [ = ]       0    .debug_frame
  +0.1%     +40  [ = ]       0    .debug_aranges
  +0.0%     +24  [ = ]       0    .debug_ranges
  +0.0%      +2  [ = ]       0    [Unmapped]
  -0.0%     -39  [ = ]       0    .strtab
  [DEL]     -76  [ = ]       0    .xt.prop._ZN4mdns7Minimal18FlatAllocatedQName19RequiredStorageSizeIJRA6_KcEEEjPS3_DpOT_
  [DEL]     -76  [ = ]       0    .xt.prop._ZN4mdns7Minimal18FlatAllocatedQName8Internal10InitializeIJRA5_KcRA6_S4_EEEvPPS4_PcS9_DpOT_
  [DEL]     -76  [ = ]       0    .xt.prop._ZN4mdns7Minimal18FlatAllocatedQName8Internal10InitializeIJRA6_KcEEEvPPS4_PcS7_DpOT_
  -0.1%     -80  [ = ]       0    .symtab
 -16.5%    -120  [ = ]       0    .xt.lit._ZN4mdns7Minimal14ResponseSenderD5Ev
  -0.0%    -124  -0.0%    -124    .flash.text
  -1.5%    -563  [ = ]       0    .shstrtab
  -0.1%   -1017  [ = ]       0    .debug_loc
  -0.1% -1.72Ki  [ = ]       0    .debug_str
  +0.0% +7.66Ki  -0.0%    -124    TOTAL
```

however I expect it may grow as we use this method more (unclear how much though, so I expect we will often use it with just 2/3/4 arguments).

Since varargs loses typesafety I think we should defer this optimization for later.

bzbarsky-apple commented:
This is code that will run on all devices, right?  I thought we wanted to generally avoid heap-allocation if we can manage it?

Is the codesize increase here expected?

andy31415 commented:
> This is code that will run on all devices, right? I thought we wanted to generally avoid heap-allocation if we can manage it?
> 
> Is the codesize increase here expected?

minMDNS code has the option to use non-heap, however the dynamic nature of the spec makes the code more readable with memory pools.

I asked some folks opinion if to try for static only (initial version, hard to read code), memory pools or Platform::Memory.

Settled for Platform::Memory because memory pools will likely need to overallocate and waste storage.

I am not convinced I trust the code size diffs (since I branched and merged with master several times).

andy31415 commented:
> > This is code that will run on all devices, right? I thought we wanted to generally avoid heap-allocation if we can manage it?
> > Is the codesize increase here expected?
> 
> minMDNS code has the option to use non-heap, however the dynamic nature of the spec makes the code more readable with memory pools.
> 
> I asked some folks opinion if to try for static only (initial version, hard to read code), memory pools or Platform::Memory.
> 
> Settled for Platform::Memory because memory pools will likely need to overallocate and waste storage.
> 
> I am not convinced I trust the code size diffs (since I branched and merged with master several times).

Ran check locallly, stats for this seem accurate: +1K code, -168 bytes BSS.
Code size change seems on the larger size for a 'just better readability'. On the other hand, the entire mdns seems to be under 10K, so it may still be ok-ish.

mspang commented:
> > > This is code that will run on all devices, right? I thought we wanted to generally avoid heap-allocation if we can manage it?
> > > Is the codesize increase here expected?
> > 
> > 
> > minMDNS code has the option to use non-heap, however the dynamic nature of the spec makes the code more readable with memory pools.
> > I asked some folks opinion if to try for static only (initial version, hard to read code), memory pools or Platform::Memory.
> > Settled for Platform::Memory because memory pools will likely need to overallocate and waste storage.
> > I am not convinced I trust the code size diffs (since I branched and merged with master several times).
> 
> Ran check locallly, stats for this seem accurate: +1K code, -168 bytes BSS.
> Code size change seems on the larger size for a 'just better readability'. On the other hand, the entire mdns seems to be under 10K, so it may still be ok-ish.

Can you try to track down some of those losses? Since code footprint is believed to be an existential risk to the project, it seems worth keeping it under control from the start. This is a fairly template heavy style even for unconstrained code; there's no logic in moving to this style if we have to undo it again after.

andy31415 commented:
Tried to re-run the bloaty diff and I get an integer overflow:

```
vscode@andreilitvin:/workspaces/connectedhomeip/examples/all-clusters-app/esp32/build$ bloaty chip-all-clusters-app.elf -- master.elf 
bloaty: integer overflow
```
happens both in VSCode image and workstation.

andy31415 commented:
> > > > This is code that will run on all devices, right? I thought we wanted to generally avoid heap-allocation if we can manage it?
> > > > Is the codesize increase here expected?
> > > 
> > > 
> > > minMDNS code has the option to use non-heap, however the dynamic nature of the spec makes the code more readable with memory pools.
> > > I asked some folks opinion if to try for static only (initial version, hard to read code), memory pools or Platform::Memory.
> > > Settled for Platform::Memory because memory pools will likely need to overallocate and waste storage.
> > > I am not convinced I trust the code size diffs (since I branched and merged with master several times).
> > 
> > 
> > Ran check locallly, stats for this seem accurate: +1K code, -168 bytes BSS.
> > Code size change seems on the larger size for a 'just better readability'. On the other hand, the entire mdns seems to be under 10K, so it may still be ok-ish.
> 
> Can you try to track down some of those losses? Since code footprint is believed to be an existential risk to the project, it seems worth keeping it under control from the start. This is a fairly template heavy style even for unconstrained code; there's no logic in moving to this style if we have to undo it again after.

Since I get 'integer overflow' on bloaty somehow, I tried just a diff:

```
> bloaty chip-all-clusters-app.elf >bloaty_malloc.txt
> bloaty master.elf >bloaty_master.txt
> diff --side-by-side bloaty*.txt         
    FILE SIZE        VM SIZE                                        FILE SIZE        VM SIZE    
 --------------  --------------                                  --------------  -------------- 
  50.2%  8.57Mi   0.0%       0    .debug_info                 |   50.2%  8.56Mi   0.0%       0    .debug_info
  17.7%  3.03Mi   0.0%       0    .debug_line                 |   17.8%  3.03Mi   0.0%       0    .debug_line
   8.9%  1.53Mi   0.0%       0    .debug_loc                       8.9%  1.53Mi   0.0%       0    .debug_loc
   8.0%  1.36Mi   0.0%       0    .debug_str                       8.0%  1.36Mi   0.0%       0    .debug_str
   4.2%   740Ki  67.6%   740Ki    .flash.text                      4.2%   740Ki  67.6%   740Ki    .flash.text
   4.1%   717Ki   0.0%       0    .debug_abbrev               |    4.1%   716Ki   0.0%       0    .debug_abbrev
   1.3%   228Ki   0.0%       0    .strtab                     |    1.3%   227Ki   0.0%       0    .strtab
   1.0%   177Ki   0.0%       0    .debug_frame                     1.0%   177Ki   0.0%       0    .debug_frame
   1.0%   168Ki   0.0%       0    .debug_ranges                    1.0%   168Ki   0.0%       0    .debug_ranges
   1.0%   168Ki  15.4%   168Ki    .flash.rodata                    1.0%   168Ki  15.4%   168Ki    .flash.rodata
   0.9%   155Ki   0.0%       0    .symtab                          0.9%   155Ki   0.0%       0    .symtab
   0.7%   113Ki  10.4%   113Ki    .iram0.text                      0.7%   113Ki  10.4%   113Ki    .iram0.text
   0.4%  67.2Ki   0.0%       0    .debug_aranges                   0.4%  67.2Ki   0.0%       0    .debug_aranges
   0.0%       0   4.0%  43.5Ki    .dram0.bss                  |    0.0%       0   4.0%  43.7Ki    .dram0.bss
   0.2%  37.7Ki   0.0%       0    .shstrtab                   |    0.2%  36.9Ki   0.0%       0    .shstrtab
   0.2%  33.2Ki   0.0%       0    [357 Others]                |    0.2%  32.8Ki   0.0%       0    [351 Others]
   0.2%  28.2Ki   2.6%  28.2Ki    .dram0.data                      0.2%  28.2Ki   2.6%  28.2Ki    .dram0.data
   0.0%  8.56Ki   0.0%       0    [Unmapped]                  |    0.1%  8.75Ki   0.0%       0    [Unmapped]
   0.0%  1.04Ki   0.1%    1024    .iram0.vectors                   0.0%  1.04Ki   0.1%    1024    .iram0.vectors
   0.0%     728   0.0%       0    .xt.lit._ZN4mdns7Minimal14R <
   0.0%     580   0.0%       0    [ELF Headers]                    0.0%     580   0.0%       0    [ELF Headers]
                                                              >    0.0%     568   0.0%       0    .xt.lit._ZN4mdns7Minimal14R
 100.0%  17.1Mi 100.0%  1.07Mi    TOTAL                          100.0%  17.1Mi 100.0%  1.07Mi    TOTAL
 ```
 
 It looks like the cost is in BSS (200 bytes?) and very little in txt (less than 1K).
 
 I think the original large diff may have been because the branch was built on top of 'operational credential advertisement so the diff contains both advertisement and memory changes. the memory alloc changes are not expected to have large impact.
 
 I would still wish I could verify though ... bloaty giving integer overflow is frustrating.

andy31415 commented:
I created a python script with the csv output of bloaty in the AdvertiserImpl.o and here is what differs in size:

```
SECTION  text
  NEW:
     28   unsigned int mdns::Minimal::FlatAllocatedQName::RequiredStorageSize<char const (&) [6]>(char const*, char const (&) [6])
     24   mdns::Minimal::QueryResponderSettings::SetReportAdditional(mdns::Minimal::FullQName const&)
     40   void mdns::Minimal::FlatAllocatedQName::Internal::Initialize<char const (&) [6]>(char const**, char*, char const*, char const (&) [6])
     22   void chip::Platform::Delete<mdns::Minimal::Responder>(mdns::Minimal::Responder*)
     42   void mdns::Minimal::FlatAllocatedQName::Internal::Initialize<char const (&) [5], char const (&) [6]>(char const**, char*, char const*, char const (&) [5], char const (&) [6])
     75   chip::Mdns::(anonymous namespace)::AdvertiserMinMdns::Clear()
     88   chip::Mdns::(anonymous namespace)::AdvertiserMinMdns::AddAllocatedResponder(mdns::Minimal::Responder*)
  REMOVED:
  DIFF:
     -214   startup._GLOBAL__sub_I__ZN4chip4Mdns17ServiceAdvertiser8InstanceEv
     705   chip::Mdns::(anonymous namespace)::AdvertiserMinMdns::Advertise(chip::Mdns::OperationalAdvertisingParameters const&)
SECTION  bss
  NEW:
  REMOVED:
  DIFF:
     -168   chip::Mdns::(anonymous namespace)::gAdvertiser
SECTION  rodata
  NEW:
     20   vtable for mdns::Minimal::Responder
     83   _ZN4chip4Mdns12_GLOBAL__N_117AdvertiserMinMdns21AddAllocatedResponderEPN4mdns7Minimal9ResponderE.str1.1
  REMOVED:
     -12   <EMPTYKEY>
  DIFF:
     -17   _GLOBAL__sub_I__ZN4chip4Mdns17ServiceAdvertiser8InstanceEv.str1.1
     119   _ZN4chip4Mdns12_GLOBAL__N_117AdvertiserMinMdns9AdvertiseERKNS0_32OperationalAdvertisingParametersE.str1.1
SECTION  literal
  NEW:
     8   unsigned int mdns::Minimal::FlatAllocatedQName::RequiredStorageSize<char const (&) [6]>(char const*, char const (&) [6])
     12   void mdns::Minimal::FlatAllocatedQName::Internal::Initialize<char const (&) [6]>(char const**, char*, char const*, char const (&) [6])
     4   void chip::Platform::Delete<mdns::Minimal::Responder>(mdns::Minimal::Responder*)
     12   void mdns::Minimal::FlatAllocatedQName::Internal::Initialize<char const (&) [5], char const (&) [6]>(char const**, char*, char const*, char const (&) [5], char const (&) [6])
     12   chip::Mdns::(anonymous namespace)::AdvertiserMinMdns::Clear()
     20   chip::Mdns::(anonymous namespace)::AdvertiserMinMdns::AddAllocatedResponder(mdns::Minimal::Responder*)
  REMOVED:
  DIFF:
     -80   startup._GLOBAL__sub_I__ZN4chip4Mdns17ServiceAdvertiser8InstanceEv
     172   chip::Mdns::(anonymous namespace)::AdvertiserMinMdns::Advertise(chip::Mdns::OperationalAdvertisingParameters const&)
SECTION  ctors
  NEW:
  REMOVED:
  DIFF:
SECTION  dtors
  NEW:
  REMOVED:
  DIFF:
  ```

andy31415 commented:
@saurabhst @BroderickCarlin @jelderton @hawk248 ?
